crates/nu-cli/src/completions/completer.rs

```diff
@@ -56,13 +57,93 @@ impl NuCompleter {
         suggestions
     }
 
+    fn external_completion(
+        &self,
+        decl_id: usize,
+        line: String,
+        pos: usize,
+        offset: usize,
+    ) -> Vec<Suggestion> {
+        let span = Span {
+            start: offset,
+            end: pos,
+        };
+
+        let line_pos = pos - offset;
+        let stack = self.stack.clone();
+
+        let block = self.engine_state.get_block(decl_id);
+        let mut callee_stack = stack.gather_captures(&block.captures);
+
+        for (idx, PositionalArg { var_id, .. }) in
+            block.signature.required_positional.iter().enumerate()
+        {
+            match idx {
+                0 => {
+                    if let Some(var_id) = var_id {
+                        callee_stack.add_var(
+                            *var_id,
+                            Value::String {
+                                val: line.clone(),
+                                span,
+                            },
+                        );
+                    }
+                }
+                1 => {
+                    if let Some(var_id) = var_id {
+                        callee_stack.add_var(
+                            *var_id,
+                            Value::Int {
+                                val: line_pos as i64,
+                                span,
+                            },
+                        );
+                    }
+                }
+                _ => {}
+            }
+        }
+
+        let result = eval_block(
+            &self.engine_state,
+            &mut callee_stack,
+            block,
+            PipelineData::new(span),
+            true,
+            true,
+        );
+
+        match result {
+            Ok(pd) => {
+                let value = pd.into_value(span);
+                match &value {
+                    Value::List { vals, span: _ } => {
+                        return map_value_completions(vals.iter(), span, offset)
+                    }
+                    _ => {}
+                }
+            }
+            Err(err) => println!("failed to eval completer block: {}", err),
```

Comment:
What should we do if we fail to eval the block?

Reply:
I'm of the idea that it should return a Result, but that will change more things in the code

Reply:
Yeah, but then it goes back to the `fn completion_helper(&mut self, line: &str, pos: usize) -> Vec<Suggestion> {` which should handle it anyway and return a `Vec<Suggestion>`